go/vt/vttablet/tabletserver/query_executor.go

```diff
@@ -506,7 +506,22 @@ func (qre *QueryExecutor) execDDL(conn *StatefulConnection) (*sqltypes.Result, e
 	}
 
 	defer func() {
-		if err := qre.tsv.se.Reload(qre.ctx); err != nil {
+		var created, altered, dropped []string
+
+		switch ddlStmt := qre.plan.FullStmt.(type) {
+		case *sqlparser.CreateTable:
+			created = append(created, ddlStmt.Table.Name.String())
+
+		case *sqlparser.AlterTable:
+			altered = append(altered, ddlStmt.Table.Name.String())
```

Comment:
Which is not what I'd expect... maybe I'm misunderstanding what those `created`, `altered` and `dropped` arrays represent?

Reply:
check if that maps contains
"users" or "\`users\`" with backticks

Reply:
Apparently `name = "users"`, without backticks, and the `tables` map doesn't have `"users"` neither with nor without the backticks.

Here's what I see on dlv:

<details><summary>🐞 Excerpt from dlv session</summary>
<p>

Running

```
dlv test vitess.io/vitess/go/vt/vttablet/tabletserver
```

I eventually got here:

```
Frame 7: ./go/vt/vttablet/tabletserver/messager/engine.go:153 (PC: 176dbe4)
   148:			mm.Close()
   149:			delete(me.managers, name)
   150:		}
   151:	
   152:		for _, name := range append(created, altered...) {
=&gt; 153:			t := tables[name]
   154:			if t.Type != schema.Message {
   155:				continue
   156:			}
   157:			if me.managers[name] != nil {
   158:				me.tsv.Stats().InternalErrors.Add("Messages", 1)
(dlv) p name
"user"
(dlv) p tables
map[string]*vitess.io/vitess/go/vt/vttablet/tabletserver/schema.Table [
	"dual": *{
		Name: (*"vitess.io/vitess/go/vt/sqlparser.IdentifierCS")(0xc002208000),
		Fields: []*vitess.io/vitess/go/vt/proto/query.Field len: 0, cap: 0, nil,
		PKColumns: []int len: 0, cap: 0, nil,
		Type: 0,
		SequenceInfo: *vitess.io/vitess/go/vt/vttablet/tabletserver/schema.SequenceInfo nil,
		MessageInfo: *vitess.io/vitess/go/vt/vttablet/tabletserver/schema.MessageInfo nil,
		CreateTime: 0,
		FileSize: 0,
		AllocatedSize: 0,},
	"test_table": *{
		Name: (*"vitess.io/vitess/go/vt/sqlparser.IdentifierCS")(0xc0016a25b0),
		Fields: []*vitess.io/vitess/go/vt/proto/query.Field len: 3, cap: 3, [
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc0017a21e0),
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc0017a2280),
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc0017a2320),
		],
		PKColumns: []int len: 1, cap: 1, [0],
		Type: 0,
		SequenceInfo: *vitess.io/vitess/go/vt/vttablet/tabletserver/schema.SequenceInfo nil,
		MessageInfo: *vitess.io/vitess/go/vt/vttablet/tabletserver/schema.MessageInfo nil,
		CreateTime: 1427325875,
		FileSize: 100,
		AllocatedSize: 150,},
	"seq": *{
		Name: (*"vitess.io/vitess/go/vt/sqlparser.IdentifierCS")(0xc0018e5c70),
		Fields: []*vitess.io/vitess/go/vt/proto/query.Field len: 4, cap: 4, [
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc0001ed180),
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc0001ed220),
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc0001ed2c0),
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc0001ed360),
		],
		PKColumns: []int len: 1, cap: 1, [0],
		Type: 1,
		SequenceInfo: *(*"vitess.io/vitess/go/vt/vttablet/tabletserver/schema.SequenceInfo")(0xc0004a5f38),
		MessageInfo: *vitess.io/vitess/go/vt/vttablet/tabletserver/schema.MessageInfo nil,
		CreateTime: 1427325875,
		FileSize: 100,
		AllocatedSize: 150,},
	"msg": *{
		Name: (*"vitess.io/vitess/go/vt/sqlparser.IdentifierCS")(0xc0016a2700),
		Fields: []*vitess.io/vitess/go/vt/proto/query.Field len: 6, cap: 6, [
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc000601800),
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc0006018a0),
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc000601940),
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc0006019e0),
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc000601a80),
			*(*"vitess.io/vitess/go/vt/proto/query.Field")(0xc000601b20),
		],
		PKColumns: []int len: 1, cap: 1, [0],
		Type: 2,
		SequenceInfo: *vitess.io/vitess/go/vt/vttablet/tabletserver/schema.SequenceInfo nil,
		MessageInfo: *(*"vitess.io/vitess/go/vt/vttablet/tabletserver/schema.MessageInfo")(0xc0016b9400),
		CreateTime: 1427325875,
		FileSize: 100,
		AllocatedSize: 150,},
]
```

</p>
</details>

Reply:
Coming back to this after a couple weeks off work...

I did some debugging today and I guess I have no idea what's going on and could use some help (maybe @harshit-gangal?). 🙂

Apparently we run DDLs 3 times for that test case, which in turn calls `reload()` and `broadcast()`, which ultimately segfaults when the optimisation is applied.

So far I was able to run this delve script against the segfaulting test both with and without the optimisation:

```
break tcase go/vt/vttablet/tabletserver/query_executor_test.go:273
cond tcase tcase.input == "create index a on user(id)"
continue
trace bcast go/vt/vttablet/tabletserver/schema/engine.go:462
on bcast print created
on bcast print altered
on bcast print dropped
break end go/vt/vttablet/tabletserver/query_executor_test.go:309
continue
quit
```

which yielded something like this with optimisation **disabled** (commented-out):

```go
created: []string len: 3, cap: 4, [
        "test_table",
        "seq",
        "msg",
]
altered: []string len: 0, cap: 0, nil
dropped: []string len: 0, cap: 0, nil

...

created: []string len: 0, cap: 0, nil
altered: []string len: 3, cap: 4, [
        "test_table",
        "seq",
        "msg",
]
dropped: []string len: 0, cap: 0, nil

...

created: []string len: 0, cap: 0, nil
altered: []string len: 3, cap: 4, [
        "test_table",
        "seq",
        "msg",
]
dropped: []string len: 0, cap: 0, nil
```

and something like this with optimisation is **enabled**:

```go
created: []string len: 3, cap: 4, [
        "test_table",
        "seq",
        "msg",
]
altered: []string len: 0, cap: 0, nil
dropped: []string len: 0, cap: 0, nil

...

created: []string len: 0, cap: 0, nil
altered: []string len: 1, cap: 1, ["user"]
dropped: []string len: 0, cap: 0, nil

SEGFAULT

```

So I'm a bit surprised to see that, when the optimisation is **disabled**, we never actually pass `"users"` in the `altered` array even though the test is supposed to alter that table. In fact, we seem to be passing the same value every time:

```go
altered: []string len: 3, cap: 4, [
        "test_table",
        "seq",
        "msg",
]
```

So I guess I don't really understand what's going on here and would really appreciate if someone could give a little hand. 🆘

Full dlv sessions here in case you wanna take a look:

<details><summary>🐞 dlv session with optimization <strong>disabled</strong></summary>

```
vscode@codespaces-ee3297:vitess$ dlv test vitess.io/vitess/go/vt/vttablet/tabletserver -- -test.run ^TestQueryExecutorPlans$
Type 'help' for list of commands.
(dlv) break tcase go/vt/vttablet/tabletserver/query_executor_test.go:273
Breakpoint tcase set at 0x18b501d for vitess.io/vitess/go/vt/vttablet/tabletserver.TestQueryExecutorPlans.func1() ./go/vt/vttablet/tabletserver/query_executor_test.go:273
(dlv) cond tcase tcase.input == "create index a on user(id)"
(dlv) continue
ERROR: logging before flag.Parse: E1024 19:56:16.704725   24914 syslogger.go:149] can't connect to syslog
> [tcase] vitess.io/vitess/go/vt/vttablet/tabletserver.TestQueryExecutorPlans.func1() ./go/vt/vttablet/tabletserver/query_executor_test.go:273 (hits goroutine(1017):1 total:1) (PC: 0x18b501d)
   268:			planWant:   "Show",
   269:			logWant:    "show create table mysql.`user`",
   270:		}}
   271:		for _, tcase := range testcases {
   272:			t.Run(tcase.input, func(t *testing.T) {
=> 273:				db := setUpQueryExecutorTest(t)
   274:				defer db.Close()
   275:				for _, dbr := range tcase.dbResponses {
   276:					db.AddQuery(dbr.query, dbr.result)
   277:				}
   278:				ctx := context.Background()
(dlv) trace bcast go/vt/vttablet/tabletserver/schema/engine.go:462
Tracepoint bcast set at 0x162522d for vitess.io/vitess/go/vt/vttablet/tabletserver/schema.(*Engine).reload() ./go/vt/vttablet/tabletserver/schema/engine.go:462
(dlv) on bcast print created
(dlv) on bcast print altered
(dlv) on bcast print dropped
(dlv) break end go/vt/vttablet/tabletserver/query_executor_test.go:309
Breakpoint end set at 0x18b5ec5 for vitess.io/vitess/go/vt/vttablet/tabletserver.TestQueryExecutorPlans.func1() ./go/vt/vttablet/tabletserver/query_executor_test.go:309
(dlv) continue
> goroutine(1017): [bcast] vitess.io/vitess/go/vt/vttablet/tabletserver/schema.(*Engine).reload(("*vitess.io/vitess/go/vt/vttablet/tabletserver/schema.Engine")(0xc000455e00), context.Context(*context.valueCtx) 0xc00067b2b0, []string len: 3, cap: 4, [...], []string len: 0, cap: 0, nil, []string len: 0, cap: 0, nil)

	created: []string len: 3, cap: 4, [
		"test_table",
		"seq",
		"msg",
	]
	altered: []string len: 0, cap: 0, nil
	dropped: []string len: 0, cap: 0, nil
> goroutine(1017): [bcast] vitess.io/vitess/go/vt/vttablet/tabletserver/schema.(*Engine).reload(("*vitess.io/vitess/go/vt/vttablet/tabletserver/schema.Engine")(0xc000455e00), context.Context(*context.emptyCtx) 0xc000cda9d8, []string len: 0, cap: 0, nil, []string len: 3, cap: 4, [...], []string len: 0, cap: 0, nil)

	created: []string len: 0, cap: 0, nil
	altered: []string len: 3, cap: 4, [
		"test_table",
		"seq",
		"msg",
	]
	dropped: []string len: 0, cap: 0, nil
> goroutine(1017): [bcast] vitess.io/vitess/go/vt/vttablet/tabletserver/schema.(*Engine).reload(("*vitess.io/vitess/go/vt/vttablet/tabletserver/schema.Engine")(0xc000455e00), context.Context(*context.emptyCtx) 0xc000cdabf8, []string len: 0, cap: 0, nil, []string len: 3, cap: 4, [...], []string len: 0, cap: 0, nil)

	created: []string len: 0, cap: 0, nil
	altered: []string len: 3, cap: 4, [
		"test_table",
		"seq",
		"msg",
	]
	dropped: []string len: 0, cap: 0, nil
> [end] vitess.io/vitess/go/vt/vttablet/tabletserver.TestQueryExecutorPlans.func1() ./go/vt/vttablet/tabletserver/query_executor_test.go:309 (hits goroutine(1017):1 total:1) (PC: 0x18b5ec5)
   304:				qre = newTestQueryExecutor(ctx, tsv, tcase.input, state.TransactionID)
   305:				got, err = qre.Execute()
   306:				require.NoError(t, err, tcase.input)
   307:				assert.Equal(t, tcase.resultWant, got, "in tx: %v", tcase.input)
   308:				assert.Equal(t, tcase.planWant, qre.logStats.PlanType, "in tx: %v", tcase.input)
=> 309:				want := tcase.logWant
   310:				if tcase.inTxWant != "" {
   311:					want = tcase.inTxWant
   312:				}
   313:				assert.Equal(t, want, qre.logStats.RewrittenSQL(), "in tx: %v", tcase.input)
   314:			})
(dlv) quit
```

</details> 


<details><summary>🐞 dlv session with optimization <strong>enabled</strong></summary>

```
vscode@codespaces-ee3297:vitess$ dlv test vitess.io/vitess/go/vt/vttablet/tabletserver -- -test.run ^TestQueryExecutorPlans$
Type 'help' for list of commands.
(dlv) break tcase go/vt/vttablet/tabletserver/query_executor_test.go:273
Breakpoint tcase set at 0x18b515d for vitess.io/vitess/go/vt/vttablet/tabletserver.TestQueryExecutorPlans.func1() ./go/vt/vttablet/tabletserver/query_executor_test.go:273
(dlv) cond tcase tcase.input == "create index a on user(id)"
(dlv) continue
ERROR: logging before flag.Parse: E1024 19:57:52.628437   25720 syslogger.go:149] can't connect to syslog
> [tcase] vitess.io/vitess/go/vt/vttablet/tabletserver.TestQueryExecutorPlans.func1() ./go/vt/vttablet/tabletserver/query_executor_test.go:273 (hits goroutine(999):1 total:1) (PC: 0x18b515d)
   268:			planWant:   "Show",
   269:			logWant:    "show create table mysql.`user`",
   270:		}}
   271:		for _, tcase := range testcases {
   272:			t.Run(tcase.input, func(t *testing.T) {
=> 273:				db := setUpQueryExecutorTest(t)
   274:				defer db.Close()
   275:				for _, dbr := range tcase.dbResponses {
   276:					db.AddQuery(dbr.query, dbr.result)
   277:				}
   278:				ctx := context.Background()
(dlv) trace bcast go/vt/vttablet/tabletserver/schema/engine.go:462
Tracepoint bcast set at 0x162522d for vitess.io/vitess/go/vt/vttablet/tabletserver/schema.(*Engine).reload() ./go/vt/vttablet/tabletserver/schema/engine.go:462
(dlv) on bcast print created
(dlv) on bcast print altered
(dlv) on bcast print dropped
(dlv) break end go/vt/vttablet/tabletserver/query_executor_test.go:309
Breakpoint end set at 0x18b6005 for vitess.io/vitess/go/vt/vttablet/tabletserver.TestQueryExecutorPlans.func1() ./go/vt/vttablet/tabletserver/query_executor_test.go:309
(dlv) continue
> goroutine(999): [bcast] vitess.io/vitess/go/vt/vttablet/tabletserver/schema.(*Engine).reload(("*vitess.io/vitess/go/vt/vttablet/tabletserver/schema.Engine")(0xc00098ea00), context.Context(*context.valueCtx) 0xc000acf2b0, []string len: 3, cap: 4, [...], []string len: 0, cap: 0, nil, []string len: 0, cap: 0, nil)

	created: []string len: 3, cap: 4, [
		"test_table",
		"seq",
		"msg",
	]
	altered: []string len: 0, cap: 0, nil
	dropped: []string len: 0, cap: 0, nil
E1024 19:57:52.820775   25720 server.go:420] Query not found: SELECT COLUMN_NAME as column_name
		FROM INFORMATION_SCHEMA.COLUMNS
		WHERE TABLE_SCHEMA = 'fakesqldb' AND TABLE_NAME = 'user'
		ORDER BY ORDINAL_POSITION
> goroutine(999): [bcast] vitess.io/vitess/go/vt/vttablet/tabletserver/schema.(*Engine).reload(("*vitess.io/vitess/go/vt/vttablet/tabletserver/schema.Engine")(0xc00098ea00), context.Context(*context.emptyCtx) 0xc00165a9c0, []string len: 0, cap: 0, nil, []string len: 1, cap: 1, [...], []string len: 0, cap: 0, nil)

	created: []string len: 0, cap: 0, nil
	altered: []string len: 1, cap: 1, ["user"]
	dropped: []string len: 0, cap: 0, nil
--- FAIL: TestQueryExecutorPlans (0.18s)
    --- FAIL: TestQueryExecutorPlans/create_index_a_on_user(id) (0.06s)
> [unrecovered-panic] runtime.fatalpanic() /usr/lib/go-1.19/src/runtime/panic.go:1143 (hits goroutine(999):1 total:1) (PC: 0x4410c0)
Warning: debugging optimized function
	runtime.curg._panic.arg: interface {}(string) "runtime error: invalid memory address or nil pointer dereference"
  1138:	// fatalpanic implements an unrecoverable panic. It is like fatalthrow, except
  1139:	// that if msgs != nil, fatalpanic also prints panic messages and decrements
  1140:	// runningPanicDefers once main is blocked from exiting.
  1141:	//
  1142:	//go:nosplit
=>1143:	func fatalpanic(msgs *_panic) {
  1144:		pc := getcallerpc()
  1145:		sp := getcallersp()
  1146:		gp := getg()
  1147:		var docrash bool
  1148:		// Switch to the system stack to avoid any stack growth, which
(dlv) quit
```

</details> 

✨✨✨